File: baml_language/sdk_tests/crates/csharp/type_shapes/customizable/roundtrip_tests/test_literals.cs
namespace CSharpParity;

using System;
using System.Collections.Generic;
using BamlSdk.Literals;

internal static class RoundTripLiterals
{
    internal static void Run()
    {
        AssertEqual(42L, Functions.ReturnLiteral42());
        AssertEqual(-1L, Functions.ReturnLiteralNegOne());
        AssertEqual("draft", Functions.ReturnLiteralDraft());
        AssertEqual("has \"quotes\"", Functions.ReturnLiteralEscaped());
        AssertEqual(true, Functions.ReturnLiteralTrue());
        AssertEqual(false, Functions.ReturnLiteralFalse());

        AssertEqual("draft", Functions.RoundTripStatus("draft"));
        AssertEqual("published", Functions.RoundTripStatus("published"));
        AssertEqual(1L, Functions.RoundTripPriority(1));
        AssertEqual(2L, Functions.RoundTripPriority(2));
    }

    private static void AssertEqual<T>(T expected, T actual)
    {
        if (!EqualityComparer<T>.Default.Equals(expected, actual))
        {
            throw new InvalidOperationException($"Expected {expected}, received {actual}.");
        }
    }
}

File: baml_language/sdk_tests/crates/swift/type_shapes/customizable/roundtrip_tests/TestLiterals.swift
// Roundtrip coverage for `Baml.literals` — port of python_pydantic2
// `roundtrip_tests/test_literals.py`.
//
// Standalone literal types collapse to their Swift base type (Swift
// has no literal types; the engine re-validates values). Literal
// *unions* get their own raw-value enums — see TestComplexModels.
import XCTest
import Baml

final class TestLiterals: XCTestCase {
    func test_literals_return_literals() throws {
        XCTAssertEqual(try Baml.literals.return_literal42(), 42)
        XCTAssertEqual(try Baml.literals.return_literal_neg_one(), -1)
        XCTAssertEqual(try Baml.literals.return_literal_draft(), "draft")
        XCTAssertEqual(try Baml.literals.return_literal_escaped(), "has \"quotes\"")
        XCTAssertEqual(try Baml.literals.return_literal_true(), true)
        XCTAssertEqual(try Baml.literals.return_literal_false(), false)
    }

    func test_literals_round_trip_literal42() throws {
        XCTAssertEqual(try Baml.literals.round_trip_literal42(x: 42), 42)
    }

    func test_literals_round_trip_literal_draft() throws {
        XCTAssertEqual(try Baml.literals.round_trip_literal_draft(x: "draft"), "draft")
    }

    func test_literals_round_trip_literal_escaped() throws {
        XCTAssertEqual(
            try Baml.literals.round_trip_literal_escaped(x: "has \"quotes\""),
            "has \"quotes\""
        )
    }

    func test_literals_round_trip_literal_true() throws {
        XCTAssertEqual(try Baml.literals.round_trip_literal_true(x: true), true)
    }

    func test_literals_round_trip_literal_false() throws {
        XCTAssertEqual(try Baml.literals.round_trip_literal_false(x: false), false)
    }

    func test_literals_round_trip_literals() throws {
        let lit = Baml.literals.Literals(
            literal_42: 42,
            literal_draft: "draft",
            literal_escaped: "has \"quotes\"",
            literal_true: true,
            literal_false: false
        )
        XCTAssertEqual(try Baml.literals.round_trip_literals(l: lit), lit)
    }
}
